### PR TITLE
keg_relocate: relocate only once

### PR DESCRIPTION
**Pull request by DoomHammer**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully run `brew tests` with your changes locally?

---

To work around a `patchelf` bug (or is it?) manifesting in either `virtual address space underrun!` or innocent looking binary being unable to execute directly (though executable through an interpreter).

### CODE REVIEW

File: Library/Homebrew/keg_relocate.rb [hunk 1]
```diff
@@ -87,15 +87,21 @@ def change_rpath(file, new_prefix)
     end
     return unless patchelf.installed?
     cmd = "#{patchelf.bin}/patchelf --set-rpath #{new_prefix}/lib"
+    test_cmd = "#{patchelf.bin}/patchelf --print-rpath --print-interpreter #{file}"
+    (old_i, old_p) = `#{test_cmd}`.split(/\n/)
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Use a string rather than a regex here `split("\n")`

File: Library/Homebrew/keg_relocate.rb [hunk 2]
```diff
@@ -87,15 +87,21 @@ def change_rpath(file, new_prefix)
...
       interpreter = new_prefix == PREFIX_PLACEHOLDER ?
         "/lib64/ld-linux-x86-64.so.2" :
         HOMEBREW_PREFIX/"lib/ld.so"
-      cmd << " --set-interpreter #{interpreter}"
+      cmd << " --set-interpreter #{interpreter}" unless old_i == interpreter
     end
     cmd << " #{file}"
-    puts "Setting RPATH of #{file}" if ARGV.debug?
-    safe_system cmd
+    if old_p == "#{new_prefix}/lib"
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: If the rpath is right but the interpreter is wrong, the interpreter won't get fixed up.
Can you change this logic to:

``` ruby
if old_p == new_prefix/"lib" && old_i == interpreter
```

File: Library/Homebrew/keg_relocate.rb [hunk 3]
```diff
@@ -87,15 +87,21 @@ def change_rpath(file, new_prefix)
     end
     return unless patchelf.installed?
     cmd = "#{patchelf.bin}/patchelf --set-rpath #{new_prefix}/lib"
+    test_cmd = "#{patchelf.bin}/patchelf --print-rpath --print-interpreter #{file}"
+    (old_i, old_p) = `#{test_cmd}`.split("\n")
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: After using backticks, check the exit value of the program with `$?`.

``` ruby
raise ErrorDuringExecution.new(cmd) unless $?.success?
```

Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: You can remove the `test_cmd` variable and merge these two lines into a single line.

**Comment by sjackman** [comment]
Looking good. One last fix.

File: Library/Homebrew/keg_relocate.rb (hunk 3 above)
Comment (DoomHammer) [inline Library/Homebrew/keg_relocate.rb]: Wanted to keep below 80ish characters per line, but if you think this is justified I'll go with it.

**Comment by sjackman** [comment]
:+1: Looks good! Thanks, Piotr.

File: Library/Homebrew/keg_relocate.rb [hunk 4]
```diff
@@ -87,15 +87,21 @@ def change_rpath(file, new_prefix)
     end
     return unless patchelf.installed?
     cmd = "#{patchelf.bin}/patchelf --set-rpath #{new_prefix}/lib"
+    (old_i, old_p) = `#{patchelf.bin}/patchelf --print-rpath --print-interpreter #{file}`.split("\n")
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: This command fails when the file is a dynamic library rather than an executable.

```
$ patchelf --print-rpath --print-interpreter /home/sjackman/.linuxbrew/Cellar/zlib/1.2.8/lib/libz.so.1.2.8
cannot find section .interp
```

Reading `--print-interpreter` will have to be moved down below inside `if file.mach_o_executable?`

File: Library/Homebrew/keg_relocate.rb [hunk 5]
```diff
@@ -87,16 +87,18 @@ def change_rpath(file, new_prefix)
     end
     return unless patchelf.installed?
     cmd = "#{patchelf.bin}/patchelf --set-rpath #{new_prefix}/lib"
-    (old_i, old_p) = `#{patchelf.bin}/patchelf --print-rpath --print-interpreter #{file}`.split("\n")
+    old_p = `#{patchelf.bin}/patchelf --print-rpath #{file}`.split("\n")
+    old_i = ""
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Delete this line, so that `old_i` is `nil` for a library.

File: Library/Homebrew/keg_relocate.rb [hunk 6]
```diff
@@ -87,16 +87,18 @@ def change_rpath(file, new_prefix)
...
     raise ErrorDuringExecution.new(cmd) unless $?.success?
     if file.mach_o_executable?
+      old_i = `#{patchelf.bin}/patchelf --print-interpreter #{file}`.split("\n")
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Add

``` ruby
raise ErrorDuringExecution.new(cmd) unless $?.success?
```

File: Library/Homebrew/keg_relocate.rb [hunk 7]
```diff
@@ -87,16 +87,18 @@ def change_rpath(file, new_prefix)
...
       interpreter = new_prefix == PREFIX_PLACEHOLDER ?
         "/lib64/ld-linux-x86-64.so.2" :
         HOMEBREW_PREFIX/"lib/ld.so"
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Change this to `"#{HOMEBREW_PREFIX}/lib/ld.so"` so that it's a string both ways, rather than a `String` and a `Pathname`.

File: Library/Homebrew/keg_relocate.rb [hunk 8]
```diff
@@ -87,16 +87,18 @@ def change_rpath(file, new_prefix)
...
       cmd << " --set-interpreter #{interpreter}" unless old_i == interpreter
     end
     cmd << " #{file}"
-    if old_p == "#{new_prefix}/lib" && old_i == interpreter
+    if old_p == "#{new_prefix}/lib" && old_i == interpreter.to_s
       puts "Skipping relocation of #{file} (RPATH already set)" if ARGV.debug?
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb:102]: Revert this change. With the above two changes, both `old_i` and `interpreter` will be `nil` for a library and a `String` for an executable, so the types are all happy and the same for comparison.

File: Library/Homebrew/keg_relocate.rb [hunk 9]
```diff
@@ -87,16 +87,18 @@ def change_rpath(file, new_prefix)
     end
     return unless patchelf.installed?
     cmd = "#{patchelf.bin}/patchelf --set-rpath #{new_prefix}/lib"
-    (old_i, old_p) = `#{patchelf.bin}/patchelf --print-rpath --print-interpreter #{file}`.split("\n")
+    old_p = `#{patchelf.bin}/patchelf --print-rpath #{file}`.split("\n")
```
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Could you please rename this to `old_rpath`?

File: Library/Homebrew/keg_relocate.rb (hunk 6 above)
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb]: Could you please rename this to `old_interpreter`?

File: Library/Homebrew/keg_relocate.rb (hunk 8 above)
Comment (sjackman) [inline Library/Homebrew/keg_relocate.rb:102]: Please rename `old_p` and `old_i` as above.

**Comment by DoomHammer** [comment]
Review taken into account.

**Comment by sjackman** [comment]
Looks great! Thanks, Piotr. Merge it and rebase #35.

**Comment by sjackman** [comment]
Thanks, Piotr!
